### PR TITLE
Loki: Search is now case insensitive

### LINKED ISSUE #15929: Explore: Loki search should be case insensitive
**What would you like to be added**:
When querying Loki for logs using a label selector and text, the text provided is searched case sensitive by default. Since the text is actually a regex, a workaround to make it case insensitive is to add a case insensitive flag: `(?i)some.*text`

**Why is this needed**:
Users may not even know the text is regex to add the flags. Or if they do, it is cumbersome for the user to know what the regex flag is to make it case insensitive. It is more intuitive for users to expect the search to be case insensitive by default. This is what other log UIs like splunk and elasticsearch do.

### PR DESCRIPTION
**Pull request by steven-sheehy**
**What this PR does / why we need it**:
See #15929

**Which issue(s) this PR fixes**:
Fixes #15929

**Special notes for your reviewer**:
- Users can still disable case insensitive search by adding `(?-i)foo`
- In the future, it might be useful to expose this option via a UI toggle. Even then though, this PR is useful since the default should still be case insensitive.

**Release note**:
```release-note
Loki now searches logs case insensitive by default
```

### CODE REVIEW

**Comment by torkelo** [comment]
@davkal any chance to review this? is this a Grafana (client) issue or better solved in on the backend?

**Review by davkal — CHANGES_REQUESTED** [review]
Grafana highlights the matched text in the browser. This uses the JS regex engine which does not support mode modifiers like case. Making search case insensitive makes sense, but highlighting searches is a behavior I like to keep.

At a minimum we can keep this change but make the highlighter work as well and accept all searches as being case insensitive.
A stretch goal could be to parse `(?-i)` and then modify the highlighter expression accordingly.

I'll also discuss with the rest of the Loki team if we want to have case insensitivity in the backend by default, but for the purpose of this PR it should be assumed that the backend stays case sensitive unless a mode modifier is sent.

File: public/app/plugins/datasource/loki/query_utils.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 const selectorRegexp = /(?:^|\s){[^{]*}/g;
+const caseInsensitive = '(?i)';
```
Comment (davkal) [inline public/app/plugins/datasource/loki/query_utils.ts]: Needs a comment that this is a mode modifier and that it's not working in JS.

**Comment by steven-sheehy** [comment]
Thanks for the feedback. I'll work on getting the highlighting to work. Wish regex worked the same between libraries/languages. 😿 

> A stretch goal could be to parse (?-i) and then modify the highlighter expression accordingly.

Might be difficult. Technically they can have any number of flags like: `(?i)(?-i)(?i)foo`

File: public/app/plugins/datasource/loki/query_utils.ts (hunk 1 above)
Comment (steven-sheehy) [reply public/app/plugins/datasource/loki/query_utils.ts]: Done

**Comment by steven-sheehy** [comment]
@davkal Added highlighting support. This handles the stretch goal of highlighting working with (?-i). It also allows `m` (muli-line mode) and `s` (single-line mode) flags as well since they are common between golang and javascript.

**Review by davkal — APPROVED** [review]

**Comment by davkal** [comment]
Works well now. Thanks for your contribution @steven-sheehy 🎉

**Pull request merged**